python/erdos/operator.py

```diff
@@ -127,6 +52,333 @@ def save_trace_events(self, file_name):
             json.dump(self._trace_events, write_file)
 
 
+class Source(BaseOperator):
+    """A :py:class:`Source` is an abstract base class that needs to be
+    inherited by user-defined source operators that generate data on a single
+    :py:class:`WriteStream` in an ERDOS dataflow graph.
+
+    A user-defined operator needs to implement the :py:func:`Source.run` and
+    :py:func:`Source.destroy` in order to take control of the execution and the
+    teardown of the operator respectively.
+    """
+    def __new__(cls, *args, **kwargs):
+        """Set up variables before call to `__init__` on the python end.
+
+        More setup is done in the Rust backend at `src/python/mod.rs`.
+        """
+        instance = super(Source, cls).__new__(cls, *args, **kwargs)
+        instance._trace_events = []
+        instance._runtime_stats = defaultdict(deque)
+        return instance
+
+    def run(self, write_stream):
+        """Runs the operator.
+
+        Invoked automatically by ERDOS, and provided with a
+        :py:class:`WriteStream` to send data on.
+
+        Args:
+            write_stream: A :py:class:`WriteStream` instance to send data on.
+        """
+        pass
```

Comment:
API question: should we force the user to implement all of these methods when writing new operators (e.g. by raising `NotImplementedError` or marking them as abstract methods?

An advantage is that the system will catch bugs, e.g. from mispelling the method name when overriding. A disadvantage is that a new operator must implement all of these methods.

Reply:
I added a logger warning if the user leaves all of the methods of an operator empty.